sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -455,6 +457,22 @@ object RemoveRedundantAliases extends Rule[LogicalPlan] {
         })
         Join(newLeft, newRight, joinType, newCondition, hint)
 
+      case u: Union =>
+        var first = true
+        plan.mapChildren { child =>
+          if (first) {
+            first = false
+            // `Union` inherits its first child's outputs. We don't remove those aliases from the
+            // first child's tree that prevent aliased attributes to appear multiple times in the
+            // `Union`'s output. A parent projection node on the top of an `Union` with non-unique
+            // output attributes could return incorrect result.
+            removeRedundantAliases(child, excluded ++ child.outputSet)
+          } else {
+            // We don't need to exclude those attributes that `Union` inherits from its first child.
```

Comment:
it's not an issue in newer spark versions because we combine adjacent Unions?

Reply:
I've no idea yet why this didn't come up in newer versions, checking now...

Reply:
Yes we did combine `Unions` in newer versions and the issue didn't show up.

Despite we didn't see extra aliases kept in 3.4, 3.3 and 3.2, shall I open a follow-up PRs with the fix commit?

Reply:
> shall I open a follow-up PRs with the fix commit?

Yes please, to make the code more defensive.

Reply:
Here is the follow-up PR: https://github.com/apache/spark/pull/37534